queue_processors/queue_processor/queueproc_utils/instrument_variable_utils.py

```diff
@@ -45,187 +51,124 @@ def log_error_and_notify(message):
-        # If we haven't found any variables, just return an empty list.
-        if variables:
-            self._update_variables(variables)
-            new_variables = []
-            for variable in variables:
-                new_variables.append(VariableUtils().copy_variable(variable))
-            return new_variables
-        return []
-
-    def get_default_variables(self, instrument_name, reduce_script=None):
+    @transaction.atomic
+    def create_variables_for_run(self, reduction_run):
         """
-        Creates and returns a list of variables from the reduction script on disk for the
-        instrument. If reduce_script is supplied, return variables using that script instead of the
-        one on disk.
+        Finds the appropriate InstrumentVariables for the given reduction run, and creates
+        RunVariables from them.
+
+        If the run is a re-run, use the previous run's variables.
+
+        If instrument variables set for the run's experiment are found, they're used.
+
+        Otherwise if variables set for the run's run number exist, they'll be used.
+
+        If not, the instrument's default variables will be used.
         """
-        if not reduce_script:
-            reduce_script = self._load_reduction_vars_script(instrument_name)
+        instrument_name = reduction_run.instrument.name
 
-        reduce_vars_module = self._read_script(reduce_script,
-                                               os.path.join(self._reduction_script_location(
-                                                   instrument_name), 'reduce_vars.py'))
+        reduce_vars_file = os.path.join(self._reduction_script_location(instrument_name),
+                                        'reduce_vars.py')
+        reduce_vars_module = self._import_module(reduce_vars_file)
+        model = self.model.variable_model
+        experiment_reference = reduction_run.experiment.reference_number
+        run_number = reduction_run.run_number
+        instrument_id = reduction_run.instrument.id
+        possible_variables = model.InstrumentVariable.objects.filter(
+            Q(experiment_reference=experiment_reference)
+            | Q(start_run__lte=run_number),
+            instrument_id=instrument_id)
+
+        variables = self.find_or_make_variables(possible_variables, run_number, instrument_id,
+                                                reduce_vars_module, 'standard_vars', False)
+        variables.extend(
+            self.find_or_make_variables(possible_variables, run_number, instrument_id,
+                                        reduce_vars_module, 'advanced_vars', True))
+
+        logger.info('Creating RunVariables')
+        # Create run variables from these instrument variables, and return them.
+        return VariableUtils().save_run_variables(variables, reduction_run)
 
-        if not reduce_vars_module:
+    def find_or_make_variables(self, possible_variables, run_number, instrument_id,
```

Comment:
- Can this take a reduction run and get the args instead from the run?
- Are the last 2 args necessary as (from what I can see within the PR) `find_or_make_variables` is only ever called twice and both within `create_variables_for_run`, meaning we could instead call this method once and have the list extension happen within the method and it returns both advanced and standard in one go?
This would reduce the number of args and be a lot clearer e.g. 
```python
def find_or_make_variables(self, possible_variables, reduction_run, reduce_vars_module):
```

Reply:
I was passing stuff as arguments to avoid db hits, but I think everything is cached anyway on the `ReductionRun` object already. In any case I've already got the variables for the filter above so I'd rather use them. We can change that if we need more fields from the ReductionRun